iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp

```diff
@@ -0,0 +1,100 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/error_handling/error_handler.hpp"
+#include "iceoryx_hoofs/testing/mocks/error_handler_mock.hpp"
+
+#include "test.hpp"
+#include <gtest/gtest-spi.h>
+
+namespace
+{
+using namespace ::testing;
+
+constexpr uint16_t MODULE_IDENTIFIER_OFFSET{42};
+
+// clang-format off
+#define TEST_ERRORS(error) \
+    error(TEST__FOOBAR) \
+    error(TEST__BARFOO)
+// clang-format on
+
+const char* TEST_ERROR_NAMES[] = {TEST_ERRORS(CREATE_ICEORYX_ERROR_STRING)};
+
+enum class KnownError : uint32_t
+{
+    NO_ERROR = iox::USER_DEFINED_MODULE_IDENTIFIER << iox::ERROR_ENUM_OFFSET_IN_BITS,
+    TEST_ERRORS(CREATE_ICEORYX_ERROR_ENUM)
+};
+
+enum class UnknownError : uint32_t
+{
+    NO_ERROR = (iox::USER_DEFINED_MODULE_IDENTIFIER + MODULE_IDENTIFIER_OFFSET) << iox::ERROR_ENUM_OFFSET_IN_BITS,
+    TEST_ERRORS(CREATE_ICEORYX_ERROR_ENUM)
+};
+
+const char* asStringLiteral(const KnownError error) noexcept
+{
+    return TEST_ERROR_NAMES[iox::errorToStringIndex(error)];
+}
+
+const char* asStringLiteral(const UnknownError error) noexcept
+{
+    return TEST_ERROR_NAMES[iox::errorToStringIndex(error)];
+}
+
+TEST(ErrorHandlerMock_test, UnsettingTemporaryErrorHandlerWithKnownModuleWorks)
+{
+    {
+        auto errorHandlerGuard = iox::ErrorHandlerMock::setTemporaryErrorHandler<KnownError>(
+            [&](const auto, const iox::ErrorLevel) { GTEST_FAIL() << "Temporary ErrorHandler shall not be called"; });
+    }
+    iox::errorHandler(KnownError::TEST__FOOBAR, iox::ErrorLevel::MODERATE);
+}
+
+TEST(ErrorHandlerMock_test, CallingErrorHandlerWithErrorOfKnownModuleAndDefaultLevelIsCaught)
+{
+    auto errorHandlerGuard =
+        iox::ErrorHandlerMock::setTemporaryErrorHandler<KnownError>([&](const auto error, const iox::ErrorLevel level) {
+            EXPECT_THAT(error, Eq(KnownError::TEST__FOOBAR));
+            EXPECT_THAT(level, Eq(iox::ErrorLevel::FATAL));
+        });
+    iox::errorHandler(KnownError::TEST__FOOBAR);
+}
+
+TEST(ErrorHandlerMock_test, CallingErrorHandlerWithErrorOfKnownModuleAndNonDefaultLevelIsCaught)
+{
+    auto errorHandlerGuard =
+        iox::ErrorHandlerMock::setTemporaryErrorHandler<KnownError>([&](const auto error, const iox::ErrorLevel level) {
+            EXPECT_THAT(error, Eq(KnownError::TEST__FOOBAR));
+            EXPECT_THAT(level, Eq(iox::ErrorLevel::MODERATE));
+        });
+    iox::errorHandler(KnownError::TEST__FOOBAR, iox::ErrorLevel::MODERATE);
```

Comment:
Same issue as in the previous case. Will succeed even when the error is not caught.